### ISSUE TITLE
Reconsider placement of management procedures

### ISSUE THREAD

**Issue by eugeneloza** [eugeneloza/decoherence#397]
As @michaliskambi suggested:

> The idea of OnBeforeRender is that it is always done before OnRender. The only difference is that FPS calculation (the "only render FPS", https://castle-engine.io/manual_optimization.php#section_fps ) does not count the time spent inside OnBeforeRender, it only counts time spent in OnRender. So **OnBeforeRender is good for some tasks that usually take 0 time**, but occasionally (e.g. on the 1st frame after loading the game) may take a long time.

- It's either about running management routines in `OnRender`,
- Or leaving it as-is (as I'm calculating FPS in a different way, without using `TCastleWindowCustom.Fps`),
- Or [_maybe better_] switch to timer-based management (as the management does not always needs to be run every frame). The problem with timer-based management is that there is no way to adjust amount of management routines to keep up with FPS which will be essential for Overworld management and might result in significantly uneven FPS. However, having pauses twice per second is still better than having FPS = 2 :) Will also require lerping some routines whenever it's possible.

Maybe running management routines is still preferable every frame and based on frame length make a decision on the amount of management used. Still unsure.

**Comment by michaliskambi** [comment]
Note that for often-occuring tasks, that however do not *have to* always occur *right before* a render, you should probably use OnUpdate, not OnBeforeRender.

OnUpdate is our general event for "things you have to constantly do", that is called "roughly as often as rendering frames", but it does not need to be synchronized with rendering so perfectly. In particular, OnUpdate is performed often even when AutoRedisplay=false and we don't render frames all the time. During it, you can look at `Fps.SecondsPassed` e.g. to make your animation time-based, or to perform something at specified time intervals. (TCastleTimer also uses Update event for this).

IOW, use OnBeforeRender only if you initialize stuff that *must* be done *right before* OnRender. I.e. 


```
procedure BeforeRender(...);
begin
  if Image = nil then
    Image := TGLImage.Create(...)
end;

procedure Render(...);
begin
  Image.Draw(...);
end;
```

For everything else, OnUpdate is probably more natural. In my games, I use OnBeforeRender almost never, while I use OnUpdate (or overridden TUIControl.Update, TCastleTransform.Update) very, very often.

**Comment by eugeneloza** [comment]
I'm not really sure if it's ok to update management in `TCastleWindow.OnUpdate`. The idea I'm trying to get is the following:
1 frame = 1/60s = |------ render -----|---- management ----|
I.e. let's imagine I have 1/120 seconds for render, so I can use up to 1/120 seconds for management and FPS will still be 60. In critical situations I can allow FPS fall down as low as 30 (if some management cannot be postponed till the next frame).
So, if I manage the world synchronously with rendering frames I have full control of FPS by increasing/decreasing amount of management. But if I use management asynchronously (`OnUpdate` fires more often than `onRender` afair, because it is also raised by e.g. `onMotion`) I don't know how much time I can spend for management without harming FPS.
This is not critical in the closed locations (dungeons) but will be extremely important in case of overworld, where the management routine is a priority-based set of nearly infinite amount of tasks (hiding/displaying/loading/unloading/changing LOD level of objects, tiles, chunks and chunks of chunks).

**Comment by michaliskambi** [comment]
If you use CastleWindow (not CastleControl) then OnUpdate is never fired from mouse motion. It's fired from TCastleApplication.UpdateAndRenderEverything (for most CastleWindow backends, including default on Windows and Linux), see here: https://github.com/castle-engine/castle-engine/blob/master/src/window/castlewindow.pas#L4851

It explicitly just calls Window.DoUpdate right before Window.DoRender :) If you have AutoRedisplay = true (so Window.Invalidated is always true) then in practice, you always get these events in simple sequence:

- OnUpdate
- OnBeforeRender
- OnRender

If you use CastleControl, then OnUpdate is *sometimes* fired from mouse motion, but that's only a workaround for Lazarus not notifying us about "idle" events often enough when you drag with mouse. Such "firing OnUpdate from OnMotion" should not occur too often (and if it does, we have a bug that we want to fix :) ). See AggressiveUpdate comments and implementation in castlecontrol.pas . If you have LimitFPS = 60, then OnUpdate will be called from TCastleControlCustom.MouseMove only to make OnUpdate calls occur 60 times per second, but not more often.

Bottom line: OnUpdate is always implemented to occur regularly, *but not more often than rendering*, because it would be wasteful to call it more often. If it was implemented differently, then the event would be useless, and it would be removed from the engine :) As you point out -- it would be dangerous to put any world processing there, if it may occur needlessly often.

**Comment by eugeneloza** [comment]
>     OnUpdate    OnBeforeRender    OnRender

Oh, that should do it! :) Thanks!
(still I'll need to update `FrameStart` time in `onRender` and it'll be fine in all cases)
